File: scripts/tasks/setup_ssh.py
"""Encrypted SSH backup and restore with atomic swap."""

from __future__ import annotations

import hashlib
import json
import os
import re
import shutil
import tempfile
from pathlib import Path
from platform import node as hostname

from scripts.common import (
    command_exists,
    ensure_dir,
    ensure_parent_dir,
    log_step,
    run_or_throw,
)
from scripts.crypto import CryptoFormat, decrypt, encrypt
from scripts.snapshot import directory_snapshot

SSH_FORMAT = CryptoFormat(magic="SCRYSSH2")


def _build_managed_block(
    github_host_alias: str,
    github_host_name: str,
    github_identity: str,
    codeberg_host_alias: str,
    codeberg_host_name: str,
    codeberg_identity: str,
) -> str:
    def block(host_alias: str, host_name: str, identity: str) -> str:
        return "\n".join(
            [
                f"Host {host_alias}",
                f"  HostName {host_name}",
                "  User git",
                f"  IdentityFile {identity}",
                "  IdentitiesOnly yes",
            ]
        )

    return "\n".join(
        [
            "# >>> scry managed git hosts >>>",
            block(github_host_alias, github_host_name, github_identity),
            "",
            block(codeberg_host_alias, codeberg_host_name, codeberg_identity),
            "# <<< scry managed git hosts <<<",
            "",
        ]
    )


def setup_ssh_backup() -> None:
    passphrase = os.environ.get("SCRY_SSH_BACKUP_PASSPHRASE", "")
    home = Path(os.environ.get("HOME", str(Path.home())))
    ssh_dir = home / ".ssh"
    repo_root = Path.cwd().resolve()
    vault_dir = repo_root / "vault" / "ssh"
    encrypted_file = Path(
        os.environ.get("SCRY_SSH_BACKUP_FILE", str(vault_dir / "ssh-keys.tar.enc"))
    )
    metadata_file = Path(
        os.environ.get("SCRY_SSH_METADATA_FILE", str(vault_dir / "ssh-keys.meta.json"))
    )

    log_step("Checking SSH backup prerequisites")
    if not command_exists("tar"):
        raise RuntimeError("Missing required tool: tar")
    print("ok: tar")

    if not ssh_dir.exists():
        raise RuntimeError(f"SSH directory not found: {ssh_dir}")

    if len(passphrase) < 16:
        raise RuntimeError(
            "Set SCRY_SSH_BACKUP_PASSPHRASE with at least 16 characters."
        )

    snap = directory_snapshot(str(ssh_dir))

    if encrypted_file.exists() and metadata_file.exists():
        try:
            metadata = json.loads(metadata_file.read_text())
            if metadata.get("sourceFingerprint") == snap.fingerprint:
                log_step("SSH backup unchanged")
                print(f"source fingerprint: {snap.fingerprint}")
                print("backup is already current; no files changed")
                return
        except (json.JSONDecodeError, KeyError):
            pass

    temp_dir = Path(tempfile.mkdtemp(prefix="scry-ssh-backup-"))

    try:
        log_step("Creating encrypted SSH archive")
        ensure_dir(vault_dir)
        ensure_parent_dir(str(encrypted_file))

        temp_tar = temp_dir / "ssh-keys.tar"
        run_or_throw(["tar", "-C", str(home), "-cf", str(temp_tar), ".ssh"])

        plaintext = temp_tar.read_bytes()
        payload = encrypt(plaintext, passphrase, SSH_FORMAT)
        encrypted_file.write_bytes(payload)
        encrypted_file.chmod(0o600)

        log_step("Writing backup metadata")
        metadata = {
            "createdAt": __import__("datetime")
            .datetime.now(__import__("datetime").timezone.utc)
            .isoformat(),
            "host": hostname(),
            "sourceDir": "~/.ssh",
            "encryptedBackupFile": str(encrypted_file.relative_to(repo_root)),
            "cipher": "aes-256-gcm",
            "kdf": "pbkdf2",
            "kdfDigest": "sha256",
            "kdfIterations": 250_000,
            "sourceFingerprint": snap.fingerprint,
            "sourceFileCount": snap.file_count,
            "sourceTotalBytes": snap.total_bytes,
            "encryptedBackupSha256": hashlib.sha256(
                encrypted_file.read_bytes()
            ).hexdigest(),
        }

        ensure_dir(metadata_file.parent)
        metadata_file.write_text(json.dumps(metadata, indent=2) + "\n")
        metadata_file.chmod(0o600)
    finally:
        shutil.rmtree(temp_dir, ignore_errors=True)

    log_step("SSH backup complete")
    print(f"created: {encrypted_file}")
    print(f"created: {metadata_file}")


def setup_ssh_restore() -> None:
    passphrase = os.environ.get("SCRY_SSH_BACKUP_PASSPHRASE", "")
    home = Path(os.environ.get("HOME", str(Path.home())))
    ssh_dir = home / ".ssh"
    repo_root = Path.cwd().resolve()
    restore_stamp = str(int(__import__("time").time() * 1000))
    staged_ssh_dir = home / f".ssh.scry-staged-{restore_stamp}"
    backup_ssh_dir = home / f".ssh.scry-backup-{restore_stamp}"

    encrypted_file = Path(
        os.environ.get(
            "SCRY_SSH_BACKUP_FILE",
            str(repo_root / "vault" / "ssh" / "ssh-keys.tar.enc"),
        )
    )
    github_host_alias = os.environ.get(
        "SCRY_GITHUB_HOST_ALIAS", "github.com-dunamismax"
    )
    github_host_name = os.environ.get("SCRY_GITHUB_HOSTNAME", "github.com")
    codeberg_host_alias = os.environ.get(
        "SCRY_CODEBERG_HOST_ALIAS", "codeberg.org-dunamismax"
    )
    codeberg_host_name = os.environ.get("SCRY_CODEBERG_HOSTNAME", "codeberg.org")
    github_identity = os.environ.get("SCRY_GITHUB_IDENTITY", "~/.ssh/id_ed25519")
    codeberg_identity = os.environ.get("SCRY_CODEBERG_IDENTITY", "~/.ssh/id_ed25519")

    log_step("Checking SSH restore prerequisites")
    if not command_exists("tar"):
        raise RuntimeError("Missing required tool: tar")
    print("ok: tar")

    if not encrypted_file.exists():
        raise RuntimeError(f"Encrypted backup not found: {encrypted_file}")

    if len(passphrase) < 16:
        raise RuntimeError(
            "Set SCRY_SSH_BACKUP_PASSPHRASE with at least 16 characters."
        )

    temp_dir = Path(tempfile.mkdtemp(prefix=f"scry-ssh-restore-{restore_stamp}-"))
    staged_ready = False
    backup_created = False
    restore_committed = False

    try:
        temp_tar = temp_dir / "ssh-keys.tar"
        extract_root = temp_dir / "extract-root"

        log_step("Decrypting and authenticating SSH archive")
        payload = encrypted_file.read_bytes()
        plaintext = decrypt(payload, passphrase, SSH_FORMAT)
        temp_tar.write_bytes(plaintext)

        log_step("Preparing staged ~/.ssh restore")
        extract_root.mkdir(parents=True)
        run_or_throw(["tar", "-C", str(extract_root), "-xf", str(temp_tar)])

        extracted_ssh = extract_root / ".ssh"
        if not extracted_ssh.exists():
            raise RuntimeError("Decrypted archive does not contain a .ssh directory.")

        if staged_ssh_dir.exists():
            shutil.rmtree(staged_ssh_dir)
        shutil.copytree(str(extracted_ssh), str(staged_ssh_dir))
        staged_ready = True

        log_step("Normalizing staged ~/.ssh permissions")

        def normalize(path: Path) -> None:
            if path.is_symlink():
                return
            if path.is_dir():
                path.chmod(0o700)
                for child in path.iterdir():
                    normalize(child)
                return
            if path.is_file():
                mode = (
                    0o644
                    if (path.suffix == ".pub" or "known_hosts" in path.name)
                    else 0o600
                )
                path.chmod(mode)

        normalize(staged_ssh_dir)

        log_step("Ensuring managed Git host entries in staged ~/.ssh/config")

        managed_start = "# >>> scry managed git hosts >>>"
        managed_end = "# <<< scry managed git hosts <<<"
        managed_block = _build_managed_block(
            github_host_alias,
            github_host_name,
            github_identity,
            codeberg_host_alias,
            codeberg_host_name,
            codeberg_identity,
        )
        staged_config = staged_ssh_dir / "config"

        existing = (
            staged_config.read_text().replace("\r\n", "\n")
            if staged_config.exists()
            else ""
        )
        pattern = re.compile(
            re.escape(managed_start) + r"[\s\S]*?" + re.escape(managed_end) + r"\n?"
        )
        without_managed = pattern.sub("", existing).strip()

        if not without_managed:
            next_config = managed_block
        else:
            next_config = f"{managed_block}\n{without_managed}"
            if not next_config.endswith("\n"):
                next_config += "\n"

        if next_config != existing:
            staged_config.write_text(next_config)
        staged_config.chmod(0o600)

        log_step("Atomically swapping staged restore into ~/.ssh")
        if ssh_dir.exists():
            ssh_dir.rename(backup_ssh_dir)
            backup_created = True

        try:
            staged_ssh_dir.rename(ssh_dir)
            staged_ready = False
            restore_committed = True
        except Exception:
            if backup_created and not ssh_dir.exists() and backup_ssh_dir.exists():
                backup_ssh_dir.rename(ssh_dir)
            raise

    except Exception:
        if backup_created and not restore_committed and not ssh_dir.exists():
            backup_ssh_dir.rename(ssh_dir)
        if staged_ready and staged_ssh_dir.exists():
            shutil.rmtree(staged_ssh_dir, ignore_errors=True)
        raise
    finally:
        shutil.rmtree(temp_dir, ignore_errors=True)

    log_step("SSH restore complete")
    print(f"restored: {ssh_dir}")
    if backup_created:
        print(f"backup: {backup_ssh_dir}")
    print(f"next: ssh -T git@{github_host_alias}")
    print(f"next: ssh -T git@{codeberg_host_alias}")

File: packages/cli/src/commands/setup-ssh.ts
import { createHash } from 'node:crypto'
import fs from 'node:fs'
import fsp from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'

import {
  decryptPayload,
  directorySnapshot,
  encryptPayload,
  ensureDir,
  ensureParentDir,
  logStep,
  runCommand,
} from '@scry-home/core'

import { commandExists } from '../lib/system'

const sshFormat = { magic: 'SCRYSSH2' } as const

const buildManagedBlock = (
  githubHostAlias: string,
  githubHostName: string,
  githubIdentity: string,
  codebergHostAlias: string,
  codebergHostName: string,
  codebergIdentity: string,
) => {
  const block = (hostAlias: string, hostName: string, identity: string) =>
    [
      `Host ${hostAlias}`,
      `  HostName ${hostName}`,
      '  User git',
      `  IdentityFile ${identity}`,
      '  IdentitiesOnly yes',
    ].join('\n')

  return [
    '# >>> scry managed git hosts >>>',
    block(githubHostAlias, githubHostName, githubIdentity),
    '',
    block(codebergHostAlias, codebergHostName, codebergIdentity),
    '# <<< scry managed git hosts <<<',
    '',
  ].join('\n')
}

export const setupSshBackup = async () => {
  const passphrase = process.env.SCRY_SSH_BACKUP_PASSPHRASE ?? ''
  const home = process.env.HOME ?? os.homedir()
  const sshDir = path.join(home, '.ssh')
  const repoRoot = process.cwd()
  const vaultDir = path.join(repoRoot, 'vault', 'ssh')
  const encryptedFile = process.env.SCRY_SSH_BACKUP_FILE ?? path.join(vaultDir, 'ssh-keys.tar.enc')
  const metadataFile =
    process.env.SCRY_SSH_METADATA_FILE ?? path.join(vaultDir, 'ssh-keys.meta.json')

  logStep('Checking SSH backup prerequisites')
  if (!commandExists('tar')) {
    throw new Error('Missing required tool: tar')
  }
  process.stdout.write('ok: tar\n')

  if (!fs.existsSync(sshDir)) {
    throw new Error(`SSH directory not found: ${sshDir}`)
  }
  if (passphrase.length < 16) {
    throw new Error('Set SCRY_SSH_BACKUP_PASSPHRASE with at least 16 characters.')
  }

  const snapshot = directorySnapshot(sshDir)
  const hasExisting =
    (await fsp
      .access(encryptedFile)
      .then(() => true)
      .catch(() => false)) &&
    (await fsp
      .access(metadataFile)
      .then(() => true)
      .catch(() => false))

  if (hasExisting) {
    try {
      const metadata = JSON.parse(await fsp.readFile(metadataFile, 'utf8')) as {
        sourceFingerprint?: string
      }
      if (metadata.sourceFingerprint === snapshot.fingerprint) {
        logStep('SSH backup unchanged')
        process.stdout.write(`source fingerprint: ${snapshot.fingerprint}\n`)
        process.stdout.write('backup is already current; no files changed\n')
        return
      }
    } catch {
      // stale metadata will be replaced
    }
  }

  const tempDir = await fsp.mkdtemp(path.join(os.tmpdir(), 'scry-ssh-backup-'))
  try {
    logStep('Creating encrypted SSH archive')
    await ensureDir(vaultDir)
    await ensureParentDir(encryptedFile)
    const tempTar = path.join(tempDir, 'ssh-keys.tar')
    await runCommand(['tar', '-C', home, '-cf', tempTar, '.ssh'])

    const plaintext = await fsp.readFile(tempTar)
    const payload = encryptPayload(plaintext, passphrase, sshFormat)
    await fsp.writeFile(encryptedFile, payload)
    await fsp.chmod(encryptedFile, 0o600)

    logStep('Writing backup metadata')
    const metadata = {
      cipher: 'aes-256-gcm',
      createdAt: new Date().toISOString(),
      encryptedBackupFile: path.relative(repoRoot, encryptedFile),
      encryptedBackupSha256: createHash('sha256')
        .update(await fsp.readFile(encryptedFile))
        .digest('hex'),
      host: os.hostname(),
      kdf: 'pbkdf2',
      kdfDigest: 'sha256',
      kdfIterations: 250_000,
      sourceDir: '~/.ssh',
      sourceFileCount: snapshot.fileCount,
      sourceFingerprint: snapshot.fingerprint,
      sourceTotalBytes: snapshot.totalBytes,
    }
    await ensureDir(path.dirname(metadataFile))
    await fsp.writeFile(metadataFile, `${JSON.stringify(metadata, null, 2)}\n`, 'utf8')
    await fsp.chmod(metadataFile, 0o600)
  } finally {
    await fsp.rm(tempDir, { force: true, recursive: true })
  }

  logStep('SSH backup complete')
  process.stdout.write(`created: ${encryptedFile}\n`)
  process.stdout.write(`created: ${metadataFile}\n`)
}

const normalizePermissions = (target: string) => {
  const stats = fs.lstatSync(target)
  if (stats.isSymbolicLink()) {
    return
  }
  if (stats.isDirectory()) {
    fs.chmodSync(target, 0o700)
    for (const child of fs.readdirSync(target)) {
      normalizePermissions(path.join(target, child))
    }
    return
  }
  if (stats.isFile()) {
    const mode =
      target.endsWith('.pub') || path.basename(target).includes('known_hosts') ? 0o644 : 0o600
    fs.chmodSync(target, mode)
  }
}

export const setupSshRestore = async () => {
  const passphrase = process.env.SCRY_SSH_BACKUP_PASSPHRASE ?? ''
  const home = process.env.HOME ?? os.homedir()
  const sshDir = path.join(home, '.ssh')
  const repoRoot = process.cwd()
  const restoreStamp = `${Date.now()}`
  const stagedSshDir = path.join(home, `.ssh.scry-staged-${restoreStamp}`)
  const backupSshDir = path.join(home, `.ssh.scry-backup-${restoreStamp}`)
  const encryptedFile =
    process.env.SCRY_SSH_BACKUP_FILE ?? path.join(repoRoot, 'vault', 'ssh', 'ssh-keys.tar.enc')
  const githubHostAlias = process.env.SCRY_GITHUB_HOST_ALIAS ?? 'github.com-dunamismax'
  const githubHostName = process.env.SCRY_GITHUB_HOSTNAME ?? 'github.com'
  const codebergHostAlias = process.env.SCRY_CODEBERG_HOST_ALIAS ?? 'codeberg.org-dunamismax'
  const codebergHostName = process.env.SCRY_CODEBERG_HOSTNAME ?? 'codeberg.org'
  const githubIdentity = process.env.SCRY_GITHUB_IDENTITY ?? '~/.ssh/id_ed25519'
  const codebergIdentity = process.env.SCRY_CODEBERG_IDENTITY ?? '~/.ssh/id_ed25519'

  logStep('Checking SSH restore prerequisites')
  if (!commandExists('tar')) {
    throw new Error('Missing required tool: tar')
  }
  process.stdout.write('ok: tar\n')
  if (!fs.existsSync(encryptedFile)) {
    throw new Error(`Encrypted backup not found: ${encryptedFile}`)
  }
  if (passphrase.length < 16) {
    throw new Error('Set SCRY_SSH_BACKUP_PASSPHRASE with at least 16 characters.')
  }

  const tempDir = await fsp.mkdtemp(path.join(os.tmpdir(), `scry-ssh-restore-${restoreStamp}-`))
  let stagedReady = false
  let backupCreated = false
  let restoreCommitted = false

  try {
    const tempTar = path.join(tempDir, 'ssh-keys.tar')
    const extractRoot = path.join(tempDir, 'extract-root')

    logStep('Decrypting and authenticating SSH archive')
    const payload = await fsp.readFile(encryptedFile)
    const plaintext = decryptPayload(payload, passphrase, sshFormat)
    await fsp.writeFile(tempTar, plaintext)

    logStep('Preparing staged ~/.ssh restore')
    await fsp.mkdir(extractRoot, { recursive: true })
    await runCommand(['tar', '-C', extractRoot, '-xf', tempTar])
    const extractedSsh = path.join(extractRoot, '.ssh')
    if (!fs.existsSync(extractedSsh)) {
      throw new Error('Decrypted archive does not contain a .ssh directory.')
    }

    await fsp.cp(extractedSsh, stagedSshDir, { recursive: true })
    stagedReady = true

    logStep('Normalizing staged ~/.ssh permissions')
    normalizePermissions(stagedSshDir)

    logStep('Ensuring managed Git host entries in staged ~/.ssh/config')
    const managedStart = '# >>> scry managed git hosts >>>'
    const managedEnd = '# <<< scry managed git hosts <<<'
    const managedBlock = buildManagedBlock(
      githubHostAlias,
      githubHostName,
      githubIdentity,
      codebergHostAlias,
      codebergHostName,
      codebergIdentity,
    )
    const stagedConfig = path.join(stagedSshDir, 'config')
    const existing = fs.existsSync(stagedConfig)
      ? (await fsp.readFile(stagedConfig, 'utf8')).replaceAll('\r\n', '\n')
      : ''
    const withoutManaged = existing
      .replace(new RegExp(`${managedStart}[\\s\\S]*?${managedEnd}\\n?`, 'g'), '')
      .trim()
    const nextConfig = withoutManaged
      ? `${managedBlock}\n${withoutManaged}${withoutManaged.endsWith('\n') ? '' : '\n'}`
      : managedBlock
    if (nextConfig !== existing) {
      await fsp.writeFile(stagedConfig, nextConfig, 'utf8')
    }
    await fsp.chmod(stagedConfig, 0o600)

    logStep('Atomically swapping staged restore into ~/.ssh')
    if (fs.existsSync(sshDir)) {
      await fsp.rename(sshDir, backupSshDir)
      backupCreated = true
    }

    try {
      await fsp.rename(stagedSshDir, sshDir)
      stagedReady = false
      restoreCommitted = true
    } catch (error) {
      if (backupCreated && !fs.existsSync(sshDir) && fs.existsSync(backupSshDir)) {
        await fsp.rename(backupSshDir, sshDir)
      }
      throw error
    }
  } catch (error) {
    if (
      backupCreated &&
      !restoreCommitted &&
      !fs.existsSync(sshDir) &&
      fs.existsSync(backupSshDir)
    ) {
      await fsp.rename(backupSshDir, sshDir)
    }
    if (stagedReady && fs.existsSync(stagedSshDir)) {
      await fsp.rm(stagedSshDir, { force: true, recursive: true })
    }
    throw error
  } finally {
    await fsp.rm(tempDir, { force: true, recursive: true })
  }

  logStep('SSH restore complete')
  process.stdout.write(`restored: ${sshDir}\n`)
  if (backupCreated) {
    process.stdout.write(`backup: ${backupSshDir}\n`)
  }
  process.stdout.write(`next: ssh -T git@${githubHostAlias}\n`)
  process.stdout.write(`next: ssh -T git@${codebergHostAlias}\n`)
}
